### PR TITLE
ICU-21810 Fix memory allocation error check in ucurr.cpp

### PR DESCRIPTION
**Pull request by FrankYFTang**
##### Checklist

- [X] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-21810
- [X] Required: The PR title must be prefixed with a JIRA Issue number. 
- [X] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [X] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [X] Issue accepted (done by Technical Committee after discussion)
- [ ] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

File: icu4c/source/common/ucurr.cpp [hunk 1]
```diff
@@ -967,22 +974,30 @@ collectCurrencyNames(const char* locale,
     CharString loc = ulocimp_getName(locale, ec2);
     if (U_FAILURE(ec2)) {
         ec = U_ILLEGAL_ARGUMENT_ERROR;
+        return;
     }
 
     // Get maximum currency name count first.
     getCurrencyNameCount(loc.data(), total_currency_name_count, total_currency_symbol_count);
 
+    *currencySymbols = nullptr;
```
Comment (roubert) [inline icu4c/source/common/ucurr.cpp:983]: I'd suggest that you move this statement to inside of the if-statement, together with setting `*total_currency_name_count` and `*total_currency_symbol_count`, that would be more logical grouping.

File: icu4c/source/common/ucurr.cpp [hunk 2]
```diff
@@ -967,22 +974,30 @@ collectCurrencyNames(const char* locale,
...
+        return;
+    }
     *currencySymbols = (CurrencyNameStruct*)uprv_malloc
         (sizeof(CurrencyNameStruct) * (*total_currency_symbol_count));
 
-    if(currencyNames == nullptr || currencySymbols == nullptr) {
-      ec = U_MEMORY_ALLOCATION_ERROR;
+    if(*currencySymbols == nullptr) {
+        *total_currency_name_count = *total_currency_symbol_count = 0;
+        uprv_free(*currencyNames);
```
Comment (roubert) [inline icu4c/source/common/ucurr.cpp:996]: You should also set this to `nullptr`.

**Review by roubert — CHANGES_REQUESTED** [review]

**Comment by FrankYFTang** [comment]
PTAL

**Review by roubert — APPROVED** [review]
LGTM

**Pull request merged**